README.md
# http hello/goodbye lab1 for CodeFellows 401

### Creator
 - Tim Combs

### Project Functionality
  - This is a Code Fellows Lab Assignment to create an http server that uses the file system for persistent storage and retrieval - it is a faux database for notes
  - The http server runs on localhost:8080

  - server.js acts as a server & router
  - dataStore.js accesses the the file system
  - readBody.js asynchronously builds the requestbody from the chunked database
  - resHandler.js writes the response and errors back to the user 

  - User will be able to read the directory of the notes, read specific notes, create new notes, update existing notes and delete notes dependent on specific path and request method
  - The dataStore.js module implements the file system methods
    - dataStore.retrieveDir implements GET requests for the /notes directory
    - dataStore.retrieveFile implements GET requests for specific notes in the directory
    - dataStore.stash implements POST requests for new specific notes for the directory
    - dataStore.update implements PUT requests to overwrite existing specific notes in the directory
    - dataStore.remove implments DELETE requests for existing specific notes in the directory
  
  - Different responses and errors will be written to the browser and/or logged to the console dependent on specific path &/or request method
  - Notes are stored as JSON files

### How To Use Codebase
  - This module uses Node, npm and the following modules:
    - net module from node, sander, http, url
    - eslint, mocha, chai for testing
  - Make sure to run npm install from the directory root to install dependencies
  - Please refer to the package.json for more info
  
  - A notes directory currently exists in the root of the project directory
  - The notes directory includes test1.json & test2.json files
    - These files do not affect the functionality of the app
    - They must be in the notes directory when starting the tests.
    - See the testing section for more details

  - To use this module as it stands, from the command line at the root of the directory type:
    ```
    $ npm start
    ``` 
  - Then open a browser window and navigate to the address localhost:8080/

  - app is a back end app, so to implement functionality project should be "wired" to a front end using the methods in the dataStore object encapsulated in dataStore.js

  - specific notes are JSON

  - testing can be done using a browser for GET requests or an app like Postman [https://www.getpostman.com/] for the other request methods


### Use Cases

  - navigating to localhost:8080/ serves index.html to the browser, which displays 'Serving pages for you using node!'

  - navigating to localhost:8080/notes displays the directory list for the /notes directory
  - navigating to localhost:8080/notes/<specific_note>.json displays the file

  - sending a POST request to localhost:8080/notes/<specific_note>.json with the note in JSON in the body writes the note into the file directory and displays 'Your file has been written'

  - sending a PUT request to localhost:8080/notes/<specific_note>.json with the note in JSON in the body overwrites the note in the file directory if it exists, creates it if it doesn't exist and displays 'Your file has been updated'

  - NOTE: PUT & POST both overwrite files if they already exist

  - sending a DELETE request to localhost:8080/notes/<specific_note>.json deletes the note from the file directory and displays 'Your file has been deleted'

  - navigating to other localhost:8080/<something_else> console.logs a 404 status code and displays 'there is no path at /<something_else> please check your map'
  

### Testing
  - Set Up
    - As stated above, both test1.json & test2.json must be in /notes folder to run tests
    - Feel free to delete for normal usage
    - test1.json
      ```
      {
        "title": "test1.json",
        "text": "Dinner is consistent, the chicken comes out golden every time"
      }
      ```

    - test2.json
      ```
      {
        "title": "test2.json",
        "text": ""
      }
      ```


### Code Shape
  - This code has been vetted using Eslint and was reviewed by Code Fellows using Travis-CI

### Collborations/Questions/issues
  - Not currently looking for collaborators at this time
  - Any questions and concerns can be handled by opening an issue on the codebase

### License
  - Licensed under the MIT license - see LICENSE.md for more info


CHANGELOG.md
# Change Log
All changes to this project will be documented in this CHANGELOG

Format is based on Keep a Changelog (http://keepachangelog.com/) and the project adheres to Semantic Versioning (http://semver.org).


## [1.2.1] - 2016-10-30
### Changed
- README.md to reflect testing
- now tests!

### Added
- 2 unit tests
- 3 E2E tests


## [1.2.0] - 2016-10-29
### Changed
- response handler split into separate file
- replaced anonymous functions with fat arrow functions in dataStore.js
- update README.md

### Added
- resHandler.js
- readBody.js 


## [1.1.0] - 2016-10-28
### Changed
- refactored API using promises
- server request listener split into separate file
- body reader split into separate file 

### Added
- index.js
- readBody.js 


## [1.0.0] - 2016-10-28
### Changed
- First release with MVP functionality

### Added
- read Directory functionality
- read File functionality
- add File functionality
- update File functionality
- delete File functionality
- CHANGELOG
- README


test/test.js
const server = require('../lib/server');

const chai = require('chai');
const chaiHttp = require('chai-http');
const expect = chai.expect;
chai.use(chaiHttp);

//unit tests on the dataStore
describe('unit testing the dataStore module', () => {
  // it('finds a directory', (done) => {
  //   chai.request(server)
  //     .get
  // });

  it('writes the body of POST request to file system', ()=> {
    chai.request(server)
      .post('/notes/test3.json')
      .set('Content-Type', 'application/json')
      .send({ title: 'test3.json', text: 'Typing has its joys' })
      .then((res) => {
        expect(res.body).to.be.equal({ title: 'test3.json', text: 'Typing has its joys' });
      })
      .catch((err) => {
        throw err;
      });
  });

  //it('reads the ')
  
  it('updates file by writing the body of a POST request', ()=> {
    chai.request(server)
      .put('/notes/test2.json')
      .send({})
      .then((res) => {
        expect(res.body).to.be.equal();
      })
      .catch((err) => {
        throw err;
      });
  });

//   it('finds a file', (done) => {
//     //expect

//   });

//   it('writes a file', (done) => {

//   });

//   it('overwrites a file', (done) => {

//   });

//   it('it deletes a file', (done) => {

//   });
});

describe('E2E testing the server', () => {
  it('returns status code = 200 on successful requests', () => {
    chai.request(server)
    .get('/')
    .then((res) => {
      expect(res).to.have.status(200);
    })
    .catch((err) => {
      throw err;
    });
  });

  it('sends response to request for specific path', () => {
    chai.request(server)
      .get('/notes')
      .then((res) => {
        //is this correct? if so, why does getting the text back tell me the server found the correct path?
        expect(res.text).to.be.equal('test1.json\ntest2.json\n');
      })
      .catch((err) => {
        throw err;
      });
  });

  it('sends response to request for specific file', () => {
    chai.request(server)
    .get('/notes/test1.json')
    .then((res) => {
      expect(res.text).to.equal('{\n  "title": "test1.json",\n  "text": "Dinner is consistent, the chicken comes out golden every time"\n}');
    })
    .catch((err) => {
      throw err;
    });
  });

  it('fails when navigating to an unknown path', () => {
    chai.request(server)
      .get('/nowhere/fast.json')
      .then((res) => {
        expect(res).to.have.status(400);
        expect(res.text).to.be.equal('that was not a valid path please check your map');
      })
      .catch((err) => {
        throw err;
      });
  });

  it('sends response to POST request for specific file', () => {
    chai.request(server)
      .post('/notes/test3.json')
      .then((res) => {
        expect(res);
      })
      .catch((err) => {
        throw err;
      });
  });

  it('sends response to PUT request for specific file', ()=> {
    chai.request(server)
      .put('/notes/test3.json')
      .then((res) => {
        expect(res);
      })
      .catch((err) => {
        throw err;
      });
  });

  it('sends response to DELETE request for specific file', () => {
    chai.request(server)
      .delete('/notes/test3.json')
      .then((res) => {
        expect(res);
      })
      .catch((err) => {
        throw err;
      });
  });


});

lib/data-store.js
const sander = require('sander');

//methods on dataStore correspond to CRUD operations
const dataStore = {};

dataStore.retrieveDir = function(fullPath, cb) {
  let relPath = '.' + fullPath;
  sander.readdir(relPath)
    .then((data) => {
      cb(null, data);
    })
    .catch(() => {
      cb({err: '404 - Directory not found'});
    });
};

//any errors thrown in a callback jump to the catch
//retrieveFile could return a promise
dataStore.retrieveFile = function(fullPath, filename, cb) {
  let relPath = '.' + fullPath;
  sander.readFile(relPath)
    .then((data) => {
      cb(null, data);
    })
    .catch(() => {
      cb({err: '404 - File not found', relPath});
    });
};

dataStore.stash = function(fullPath, body, cb) {
  let relPath = '.' + fullPath;
  sander.writeFile(relPath, body)
    .then(() => {
      cb(null, {text: 'Your file has been written'});
    })
    .catch(() => {
      cb({err: '500 - Internal Server Error'});
    });
};

dataStore.update = function(fullPath, body, cb) {
  let relPath = '.' + fullPath;
  sander.writeFile(relPath, body)
    .then(() => {
      cb(null, {text: 'Your file has been written'});
    })
    .catch(() => {
      cb({err: '500 - Internal Server Error'});
    });
};

dataStore.remove = function(fullPath, cb) {
  let relPath = '.' + fullPath;
  sander.unlink(relPath)
    .then(() => {
      cb(null, {text: 'Your file has been deleted'});
    })
    .catch(() => {
      cb({err: '500 - Internal Server Error'});
    });
};








module.exports = dataStore;

lib/resHandler.js
const resHandler = {};

resHandler.writeDir = (data, res) => {
  if (data.length === 0) {
    res.write('There are no notes, add some!');
  }else{
    data.forEach((file) => {
      res.write(file);
      res.write('\n');
    });
  }
  res.end();
};

resHandler.writeFile = (data, res) => {
  //if no data in the note part
  if (!JSON.parse(data)['text']) {
    res.write('That note is empty. Perhaps you meant to update the note?');
  }else{
    res.write(data);
  }
  res.end();
};

resHandler.writeDeleteMsg = (res) => {
  res.write('Your file has been deleted');
  res.end();
};

resHandler.writeUpdateMsg = (res) => {
  res.write('Your file has been updated');
  res.end();
};

//msg created in dataStore
resHandler.writeStashMsg = (res) => {
  res.write('Your file has been stashed!');
  res.end();
};

resHandler.errHandler = (res) => {
  let status = res.statusCode;
  let msg = '';
  if (status === 400) {
    msg = 'that was not a valid path please check your map';
  }else if (status === 404) {
    msg = 'file or directory not found';
  }else if (status === 410) {
    msg = 'That is not a valid note. Perhaps you meant to create a new note?';
  }else if (status === 500) {
    msg = 'our server is having difficulties, sorry';
  }
  res.write(msg);
  res.end();
};
  
module.exports = resHandler;

lib/server.js
const http = require('http');
const url = require('url');
const sander = require('sander');

const readBody = require('./readBody.js');
const dataStore = require('./dataStore.js');
const resHandler = require('./resHandler.js');

const indexHtml = sander.createReadStream('index.html');

const server = http.createServer((req, res) => {
  res.statusCode = 200;

  const parsedUrl = url.parse(req.url, true);
  //const queryObj = parsedUrl.query;
  const httpVerb = req.method;

  const fullPath = parsedUrl.pathname;
  const relPath = '.' + fullPath;
  const splitPath = fullPath.split('/');

  const dirName = '/' + splitPath[1];
  const relDirName = '.' + dirName;
  const filename = splitPath[2];

  //server decides what CRUD method then dataStore handles the fs operation
  //why does not serve the index.html page?
  //is it serving it and reloading?
  if (dirName === '/') {
    indexHtml.pipe(res);
  }else if(dirName === '/notes' && (filename) && httpVerb === 'POST') {
    readBody(req)
      .then((body) => {
        dataStore.stash(relPath, body, filename, relDirName);
      })
      .then((data) => {
        resHandler.writeStashMsg(res, data);
      })
    .catch((err) => {
      res.statusCode = 500;
      resHandler.errHandler(res);
      console.log(err);
    });
  }else if(dirName === '/notes' && (filename) && httpVerb === 'PUT') {
    readBody(req) 
      .then((body) => {
        dataStore.update(relPath, body);
      })
      .then(() => {
        resHandler.writeUpdateMsg(res);
      })
      .catch((err) => {
        res.statusCode = 500;
        resHandler.errHandler(res);
        console.log(err);
      });
  }else if(dirName === '/notes' && httpVerb === 'DELETE') {
    dataStore.remove(relPath)
      .then(() => {
        resHandler.writeDeleteMsg(res);
      })
      .catch((err) => {
        res.statusCode = 404;
        resHandler.errHandler(res);
        console.log(err);
      });
  }else if(dirName === '/notes' && (filename) && httpVerb === 'GET') {
    dataStore.retrieveFile(relPath)
      .then((data) => {
        resHandler.writeFile(data, res);
      })
      .catch((err, data) => {
        if (!data) {
          res.statusCode = 410;
          resHandler.errHandler(res);
          console.log(err);
        }else{
          res.statusCode = 500;
          resHandler.errHandler(res);
          console.log(err);
        }
        res.end();
      });
  }else if(dirName === '/notes' && httpVerb === 'GET') {
    dataStore.retrieveDir(relPath)
      .then((data) => {
        resHandler.writeDir(data, res);
      })
      .catch((err) => {
        res.statusCode = 500;
        resHandler.errHandler(res);
        console.log(err);
      });
  }else{
    res.statusCode = 400;
    resHandler.errHandler(res);
    console.log(res.statusCode, 'not valid path');
  }
});

module.exports = server;


//use promises in place of callbacks in the server - affects dataStore.js
//take functionality inside each case in server and encapsulate into separate functions
//revisit separated case functions and maybe make them methods on a object that is in a separate file
//differntiate functionality for put & post